Re-institute purging duplicate NEVRA after an rpm repo sync

seandst opened the pull request:
https://pulp.plan.io/issues/1461
fixes #1461

See the redmine issues for a rude but effective way to generate a few dozen thousand duplicate packages to stress-test the duplicate nevra generator.

seandst commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +288,104 @@ def remove_unit_duplicate_nevra(unit, repo):
+    """
+    # In order to find potential duplicate nevras, a mongo aggregation pipeline is employed
+    # to sort a unit's entire collection based on non-checksum unit key fields. Doing the sorting
+    # purely as a normal QuerySet overflow's mongo's sort stage bugger, and doing it purely with
```
This typo is too funny to fix. :D

seandst commented:
rhel6 is failing because mongoengine adds a kwarg to the aggregation pipeline that I neither want want nor use which is unsupported in that version of mongo. :(

rbarlow commented on extensions_admin/pulp_rpm/extensions/admin/status.py:
```diff
@@ -330,3 +333,18 @@ def update_func(new_state):
+        # Example Data:
+        # "purge_duplicates": {
+        #   "state": "FINISHED"
+        # }
```
Can you turn the above into a docblock?

seandst replied:
This is [a foolish consistency](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds), and looks like the docblocks of all similar methods in this module as a result.

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +287,106 @@ def remove_unit_duplicate_nevra(unit, repo):
+      }
+      return collector
+    }
+    """
```
Haha, someone just wrote pulp-list today to ask if we use the MongoDB JS engine. So I guess you can write back with a "we do nooooow!" or similar. ☺

seandst replied:
We really shouldn't, though. :(

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +287,106 @@ def remove_unit_duplicate_nevra(unit, repo):
+    """duplicate NEVRA unit ID generator
+
+    :param unit: The unit whose NEVRA should be removed
+    :type unit: ContentUnit
```
We usually put these at the end of the docblock, but I don't know if that matters.

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +287,106 @@ def remove_unit_duplicate_nevra(unit, repo):
+    else:
+        name_field = 'name'
+
+    # going with % interpolation here because all the {}'s make .format no fun for JS
```
s/JS/JS./

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +287,106 @@ def remove_unit_duplicate_nevra(unit, repo):
+    function (key, values) {
+      // collect mapped values into the first value to build the list of ids for this key/nevra
+      var collector = values[0]
+      // since collector is values[0] start this loop at index 1
```
s/1/1./

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +287,106 @@ def remove_unit_duplicate_nevra(unit, repo):
+    for reduced in reduced_units:
+        # filter out the undefined (now None) values created in the finalize step
+        if reduced.value:
+            yield reduced.value['ids']
```
I like this solution, an elegant approach to the problem.

rbarlow commented:
I think we can ignore the Jenkins failures, as they appear to be environmental. Nice work, and thanks for such good inline comments explaining something a little unusual for Pulp code. I also like that the three commits are each atomic, so no need to squash if you don't want to.

seandst commented:
After some discussion, both the mongo 2.6+ aggregation mechanism and the slower mapreduce method will be used, depending on the available version of mongo. I've removed the LGTM label since this will need to be reviewed again once I've updated it.

seandst commented:
After a fair bit of fist-shaking aimed at el6, this PR has been updated. Comments welcome!

seandst commented:
ok test

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +289,209 @@ def remove_unit_duplicate_nevra(unit, repo):
+        unit.objects.aggregate({'$limit': 1})
+        return _duplicate_key_id_generator_aggregation(unit, fields)
+    except OperationFailure:
+        # mongodb doesn't support aggregation cursors, use the slower mapreduce method
```
s/cursors/cursors (womp womp)/

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +289,209 @@ def remove_unit_duplicate_nevra(unit, repo):
+        # mongodb doesn't support aggregation cursors, use the slower mapreduce method
+        _logger.info('Purging duplicate NEVRA can take significantly longer in versions of '
+                     'mongodb lower than 2.6. Consider upgrading mongodb if cleaning duplicate '
+                     'packages takes an unreasonably long time.')
```
We usually gettext our human readable messages, despite not having any translation files.

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +289,209 @@ def remove_unit_duplicate_nevra(unit, repo):
+                # this iteration is the first to detect a duplicate for this nevra.
+                # Initialize the list of ids to yield with the *previous* id,
+                # since it was the first unit seen with the current nevra
+                yielding_ids = [previous_pkg_id]
```
What if you intialized yielding_ids to [] instead of None so you can always append below? Of course, you'd also need to reset it to [] when moving on to the next nevra, but it might make the code slightly smaller and let you have less 'splainin' to do.

rbarlow commented on plugins/pulp_rpm/plugins/importers/yum/purge.py:
```diff
@@ -286,3 +289,209 @@ def remove_unit_duplicate_nevra(unit, repo):
+    # after the pkg loop, if the last pkg was a duplicate, the yielding else clause won't run
+    # in that case, do one final yield to make sure all potential duplicates are yielded
+    if yielding_ids:
+        yield yielding_ids
```
Clever.

rbarlow commented:
I like the aggregate solution, nicely done!

rbarlow commented:
It seems that Jenkies is just busted, so as what seems to be usual lately I think we can ignore it.
